File: dfs/841_keys_and_rooms.py
class Solution:
    # O(V+E) time | O(V) space
    # where V is the number of vertices and E is the number of edges
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visit = set()

        def dfs(curRoom):
            visit.add(curRoom)
            for nextRoom in rooms[curRoom]:
                if nextRoom not in visit:
                    dfs(nextRoom)

        dfs(0)
        return len(visit) == len(rooms)

File: dfs/841_keys_and_rooms.go
// O(V+E) time | O(V) space 
// V is the number of vertices and E is the number of edges
func canVisitAllRooms(rooms [][]int) bool {
    visit := make(map[int]bool)
    dfs(0, visit, rooms)
    return len(visit) == len(rooms)
}

func dfs(room int, visit map[int]bool, rooms [][]int) {
    if _, ok := visit[room]; ok {
        return 
    }
    visit[room] = true
    for _, nextRoom := range rooms[room] {
        dfs(nextRoom, visit, rooms)
    }
}
